File: sample-projects/objectivec/CellarsAndCentaurs/Adapter/Persistence/Repository.m
#import "Repository.h"

@interface Repository ()

@property (nonatomic, strong) NSMutableDictionary<NSString *, id> *storage;

@end

@implementation Repository

- (instancetype)init {
    self = [super init];
    if (self) {
        _storage = [NSMutableDictionary dictionary];
    }
    return self;
}

- (void)save:(id)object withIdentifier:(NSString *)identifier {
    self.storage[identifier] = object;
}

- (id)findOne:(NSString *)identifier {
    return self.storage[identifier];
}

- (NSArray *)findAll {
    return self.storage.allValues;
}

@end

File: training/swift/Sources/CellarsAndCentaurs/Adapter/Persistence/Repository.swift
class Repository<T: Identifiable> where T.ID == String {
    private var storage: [String: T] = [:]

    func save(_ item: T) {
        storage[item.id] = item
    }

    func findOne(_ id: String) -> T? {
        storage[id]
    }

    func findAll() -> [T] {
        Array(storage.values)
    }
}
